### PR TITLE
feat: Add @link directive support for schema composition

### LINKED ISSUE #2008: feat: Implement @link directive for schema composition (Link Spec v1.0)
# Implement @link directive support for schema composition

## Overview
Implement support for the @link directive as defined in the [Apollo Link Spec v1.0](https://specs.apollo.dev/link/v1.0/). This directive enables schema composition and linking, replacing our custom `tanka_import` syntax.

## Specification
The @link directive allows schemas to reference and import types/directives from other schemas:

```graphql
extend schema @link(
  url: "https://specs.apollo.dev/federation/v2.0",
  import: ["@key", "@external", "FieldSet"]
)
```

#### ISchemaLoader Interface
```csharp
public interface ISchemaLoader
{
    Task<TypeSystemDocument?> LoadSchemaAsync(string url, CancellationToken cancellationToken = default);
    bool CanLoad(string url);
}
```

**Built-in Implementations:**
- `FileSchemaLoader` - Load from file:// URLs and relative paths
- `HttpSchemaLoader` - Load from HTTP/HTTPS URLs  
- `CompositeSchemaLoader` - Chain multiple loaders

#### SchemaBuildOptions Updates
```csharp
public class SchemaBuildOptions
{
    /// <summary>
    /// Schema loader for processing @link directives. 
    /// Setting this automatically enables link processing.
    /// </summary>
    public ISchemaLoader? SchemaLoader { get; set; }
    
    /// <summary>
    /// Whether to process @link directives. 
    /// Automatically true when SchemaLoader is provided.
    /// </summary>
    public bool ProcessLinkDirectives => SchemaLoader \!= null;
    
    /// <summary>
    /// Maximum depth for recursive @link processing (default: 10)
    /// </summary>
    public int MaxLinkDepth { get; set; } = 10;
}
```

### 3. Processing Algorithm

**Depth-First Processing:**
1. Load schema from URL
2. Extract @link directives from schema
3. Recursively process each linked schema (depth-first)
4. Apply import filtering based on LinkInfo
5. Merge schemas in dependency order (dependencies first, then dependent schema)

**Key Methods:**
- `ProcessLinkDirectivesAsync()` - Main entry point in Build()
- `LoadAndResolveSchemaAsync()` - Recursive depth-first loader
- `ApplyImportFiltering()` - Filter types/directives based on imports

### 4. Import Filtering

Support for selective imports with aliasing:
```graphql
# Import specific items
@link(url: "./types.graphql", import: ["User", "Post"])

# Import with aliasing  
@link(url: "./types.graphql", import: [{name: "User", as: "MyUser"}])

# Import directives
@link(url: "./directives.graphql", import: ["@auth", "@deprecated"])
```

### 5. URL Support

Multiple URL schemes supported:
- `https://example.com/schema.graphql` - HTTP(S) URLs
- `file://./schemas/types.graphql` - File URLs
- `./types/user.graphql` - Relative file paths
- Future: `resource://Assembly/schema.graphql` - Embedded resources

### 6. Error Handling

- Circular dependency detection via visited URL tracking
- Maximum depth enforcement to prevent infinite recursion
- Clear error messages for:
  - Network failures
  - File not found
  - Parse errors
  - Missing imports
  - Circular dependencies

## Testing Requirements

1. **Unit Tests:**
   - Each ISchemaLoader implementation
   - Import filtering logic
   - Circular dependency detection
   - Max depth enforcement

2. **Integration Tests:**
   - End-to-end @link processing
   - Multi-level schema linking
   - Import aliasing
   - Error scenarios

## Migration from tanka_import

The custom `tanka_import` syntax has been completely removed. Users should migrate to @link directive:

**Before:**
```graphql
tanka_import "./types/user.graphql"
```

**After:**
```graphql
extend schema @link(url: "./types/user.graphql")
```

## Related Work
- PR removing tanka_import syntax (completed)
- Apollo Federation v2 support (#2009) will use this for federation specs

## Success Criteria
- [x] @link directive defined in core schema
- [x] LinkDirectiveProcessor for extracting link information
- [ ] ISchemaLoader interface and implementations
- [ ] Depth-first recursive schema loading
- [ ] Import filtering with aliasing support
- [ ] Comprehensive test coverage
- [ ] Documentation updates

### PR DESCRIPTION
**Pull request by pekkah**
## Summary

Implement support for @link directive as defined in [Apollo Link Spec v1.0](https://specs.apollo.dev/link/v1.0/), enabling schema composition and linking. This replaces the custom `tanka_import` syntax with a standard GraphQL approach.

### 📦 Pluggable Schema Loading
- **ISchemaLoader interface** with built-in implementations:
  - `FileSchemaLoader` - Load from file system and file:// URLs
  - `HttpSchemaLoader` - Load from HTTP/HTTPS URLs  
  - `CompositeSchemaLoader` - Chain multiple loaders

### 🔄 Smart Schema Processing
- **Depth-first resolution** - Dependencies loaded before dependents
- **Circular dependency detection** - Prevents infinite loops
- **Maximum depth enforcement** - Configurable limit (default: 10)
- **Duplicate prevention** - Automatically filters existing types

### 🎯 Import Filtering
```graphql
# Import specific types
@link(url: "./types.graphql", import: ["User", "Post"])

# Import directives
@link(url: "./federation.graphql", import: ["@key", "@external"])

# Import everything (default)
@link(url: "./schema.graphql")
```

### ⚡ Automatic Processing
- Link processing enabled automatically when `SchemaLoader` is provided
- Clean, intuitive API: just set `options.SchemaLoader`

## 🚨 Breaking Changes

**Removed custom `tanka_import` syntax completely**

**Migration:**
```graphql
# Before
tanka_import "./types.graphql"

# After  
extend schema @link(url: "./types.graphql")
```

### Built-in Types Extended
```graphql
directive @link(
  url: String\!,
  as: String,
  import: [link__Import],
  for: link__Purpose
) repeatable on SCHEMA

scalar link__Import
enum link__Purpose { SECURITY EXECUTION }
```

### Usage Example
```csharp
var builder = new SchemaBuilder();
builder.Add(mainSchema);

var schema = await builder.Build(new SchemaBuildOptions
{
    SchemaLoader = CompositeSchemaLoader.CreateDefault(),
    Resolvers = resolvers
});
```

## 🧪 Testing

- **20 comprehensive tests** covering:
  - All schema loader implementations
  - Complex multi-level linking scenarios
  - Circular dependency handling
  - Import filtering and edge cases
  - Error conditions and validation

## 📋 Test Coverage

| Component | Tests | Coverage |
|-----------|-------|----------|
| SchemaLoaders | 8 tests | All implementations |
| Link Processing | 8 tests | End-to-end scenarios |
| Link Parsing | 4 tests | Directive processing |

## 🔗 Related Issues

- Resolves #2008 - Implement @link directive support
- Foundation for #2009 - Apollo Federation v2 update

## 📚 References

- [Apollo Link Spec v1.0](https://specs.apollo.dev/link/v1.0/)
- [Apollo Federation v2 Docs](https://www.apollographql.com/docs/federation/)

---

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: src/GraphQL/TypeSystem/HttpSchemaLoader.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+using Tanka.GraphQL.Language;
+using Tanka.GraphQL.Language.Nodes.TypeSystem;
+
+namespace Tanka.GraphQL.TypeSystem;
+
+/// <summary>
+/// Loads GraphQL schemas from HTTP/HTTPS URLs
+/// </summary>
+public class HttpSchemaLoader : ISchemaLoader
```
[the remark this answers is not in the document]
Comment (pekkah) [reply src/GraphQL/TypeSystem/HttpSchemaLoader.cs]: ✅ **Fixed**: HttpSchemaLoader now implements IDisposable and properly disposes the HttpClient when owned. CompositeSchemaLoader also implements IDisposable to dispose nested disposable loaders. See commit 32eabe61.

File: src/GraphQL/TypeSystem/HttpSchemaLoader.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            var content = await response.Content.ReadAsStringAsync(cancellationToken);
+
+            var parser = Parser.Create(content);
+            return parser.ParseTypeSystemDocument();
+        }
+        catch (Exception)
+        {
+            // Log error if logging is available
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (pekkah) [reply src/GraphQL/TypeSystem/HttpSchemaLoader.cs]: ✅ **Fixed**: Updated HttpSchemaLoader to catch specific exceptions (HttpRequestException, TaskCanceledException, InvalidOperationException) instead of generic Exception. This provides better error diagnostics. See commit 32eabe61.

File: src/GraphQL/TypeSystem/LinkDirectiveProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            {
+                if (extension.Directives != null)
+                {
+                    links.AddRange(ProcessDirectives(extension.Directives));
+                }
+            }
+        }
+
+        return links;
+    }
```
[the remark this answers is not in the document]
Comment (pekkah) [reply src/GraphQL/TypeSystem/LinkDirectiveProcessor.cs]: ✅ **Fixed**: This is the same as the earlier comment - converted LinkDirectiveProcessor.ProcessLinkDirectives to use yield return instead of List allocation for better memory efficiency and lazy evaluation. See commit 32eabe61.
